app/server/lib/ActiveDoc.ts

```
@@ -2590,6 +2598,33 @@ export class ActiveDoc extends EventEmitter {
     }];
   }
 
+  private async _updateAttachmentFileSizesUsingFileIdent(docSession: OptDocSession,
+                                                         newFileSizesByFileIdent: Map<string, number>): Promise<void> {
+    const rowIdsToUpdate: number[] = [];
+    const newFileSizesForRows: CellValue[] = [];
+    const attachments = this.docData?.getMetaTable("_grist_Attachments").getRecords();
+    for (const attachmentRec of attachments ?? []) {
+      const newSize = newFileSizesByFileIdent.get(attachmentRec.fileIdent);
+      if (newSize && newSize !== attachmentRec.fileSize) {
+        rowIdsToUpdate.push(attachmentRec.id);
+        newFileSizesForRows.push(newSize);
+      }
+    }
+
+    const action: BulkUpdateRecord = ['BulkUpdateRecord', '_grist_Attachments', rowIdsToUpdate, {
+      fileSize: newFileSizesForRows
+    }];
+
+    await this._applyUserActionsWithExtendedOptions(
```

What do you think about wrapping it with `try catch` ? Asking because, this will throw if ids are not valid, it is very unlikely but still possible if the docData we have in memory is not exactly the same as the one in data engine, everything here is async, so maybe it can somehow interfere with other modifications. 


Ideally this would all be done in "transaction like" scope, but I don't know if it is important to worry about that here. We have some mechanism for locking the doc modifications, but maybe this is overkill here.  But there is other mechanism available - the upsert action `BulkAddOrUpdateRecord`, where you can update those values based on the fileIdentity, not the ids.

---

This is a difficult question. The options seem to be:
- Don't catch, fail to import the attachments
- Catch, import the attachments with an incorrect file size

If this is very rare, I think I'd prefer to let it fail and force the user to re-import, than import attachments with an invalid file size.

I don't think upsert is an option unfortunately, as it's not correct to add a new record if it doesn't exist. Best case, we have an extra attachment being tracked that doesn't exist. Worst case, it breaks other code because it's not a fully valid record that's added (unless we try to reinsert the whole record...).


My inclination is to leave it, add a comment, see if it actually comes up in production.

How do other parts of the codebase handle the in-memory data and data engine being out of sync? Just locking?

---

The minimum was to wrap this in try catch, and rethrow something more meaningful, explaining what has happen.

I don't know for sure, but we have some data corruptions problems we don't know how to fix or don't know the reason. So here, we know that the ids can go wrong, so we can anticipate it and explain in more details.

Since we don't do any performance or real load tests, waiting for this to happen on prod is a wrong idea, we don't know anything about how this would work on prod.

Regarding the `BulkAddOrUpdateRecord`, there is a on option parameter that disables adding if it can't find records to update, so it does look as a good option.